File: solutions/java/knapsack/1/src/main/java/Knapsack.java
import java.util.List;

class Knapsack {

    int maximumValue(int maximumWeight, List<Item> items) {
        // dp[w] = maximum value achievable with weight capacity w
        int[] dp = new int[maximumWeight + 1];
        
        // Process each item exactly once (0/1 knapsack)
        for (Item item : items) {
            // Iterate BACKWARDS to avoid reusing the same item
            for (int w = maximumWeight; w >= item.weight; w--) {
                // Choose: take item or skip it — whichever gives more value
                dp[w] = Math.max(dp[w], dp[w - item.weight] + item.value);
            }
        }
        
        return dp[maximumWeight];
    }
}

File: solutions/javascript/knapsack/1/knapsack.js
export const knapsack = (maximumWeight, items) => {
  // Create DP table: dp[w] = maximum value achievable with weight capacity w
  const dp = new Array(maximumWeight + 1).fill(0);
  
  for (const item of items) {
    // Iterate backwards to avoid reusing the same item multiple times
    for (let w = maximumWeight; w >= item.weight; w--) {
      const includeItem = dp[w - item.weight] + item.value;
      const excludeItem = dp[w];
      dp[w] = Math.max(includeItem, excludeItem);
    }
  }
  
  return dp[maximumWeight];
};
